instrumentation/kafka/src/test/java/brave/kafka/ITTracingKafka.java

```diff
@@ -0,0 +1,161 @@
+
+    @Before
+    public void setUp() throws Exception {
+        consumerTracing = Tracing.newBuilder()
+            .reporter(consumerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+        producerTracing = Tracing.newBuilder()
+            .reporter(producerSpans::add)
+            .sampler(Sampler.ALWAYS_SAMPLE)
+            .build();
+
+        initKafkaServer();
+    }
+
+    @After
+    public void closeKafkaServer() {
+        kafkaServer.shutdown();
+    }
+
+    @Test
+    public void produce_and_consume_kafka_message() throws Exception {
+        TracingProducer<String, String> tracingProducer = createTracingProducer();
+        TracingConsumer<String, String> tracingConsumer = createTracingConsumer();
+
+        Future<RecordMetadata> send = tracingProducer.send(new ProducerRecord<>(TEST_TOPIC, TEST_KEY, TEST_VALUE));
+        // Block for synchronous send
+        send.get();
+
+        ConsumerRecords<String, String> records = tracingConsumer.poll(1000);
+        // Oddly we need to poll twice to retrieve content
+        if (records.isEmpty()) {
+            records = tracingConsumer.poll(1000);
+        }
+
+        assertThat(records).hasSize(1);
+        assertThat(producerSpans).hasSize(1);
+        assertThat(consumerSpans).hasSize(1);
+
+        assertThat(Long.toHexString(consumerSpans.getFirst().traceId))
+            .isEqualTo(Long.toHexString(producerSpans.getFirst().traceId));
+
+        RecordTracing recordTracing = new RecordTracing(consumerTracing);
+        for (ConsumerRecord<String, String> record : records) {
+            brave.Span span = recordTracing.nexSpanFromRecord(record);
+            assertThat(span.context().parentId()).isEqualTo(producerSpans.getLast().traceId);
+        }
+    }
+
+    private void initKafkaServer() throws IOException {
+        String zkAddress = setupZookeeper();
+        setupKafka(zkAddress);
+    }
+
+    private String setupZookeeper() {
+        EmbeddedZookeeper zkServer = new EmbeddedZookeeper();
+        String zkAddress = LOCALHOST + ":" + zkServer.port();
+        ZkClient zkClient = new ZkClient(zkAddress, 30000, 30000, ZKStringSerializer$.MODULE$);
+        zkUtils = ZkUtils.apply(zkClient, false);
+        return zkAddress;
+    }
+
+    private void setupKafka(String zkAddress) throws IOException {
+        Properties brokerProps = new Properties();
+        brokerProps.setProperty("zookeeper.connect", zkAddress);
+        brokerProps.setProperty("broker.id", "0");
+        brokerProps.setProperty("log.dirs", Files.createTempDirectory("kafka-").toAbsolutePath().toString());
+        brokerProps.setProperty("listeners", "PLAINTEXT://" + LOCALHOST + ":" + KAFKA_PORT);
+
+        brokerProps.setProperty("replica.socket.timeout.ms", "1000");
+        brokerProps.setProperty("controller.socket.timeout.ms", "1000");
+        brokerProps.setProperty("offsets.topic.replication.factor", "1");
```

Comment:
I don't think it is contributing to the intermittent test failure, but you could reduce churn on broker startup by also reducing the number of partitions for the consumer offsets topic from the default of 50 to 1.

`brokerProps.setProperty("offsets.topic.num.partitions", "1");`